File: internal/plugins/jest/rules/no_duplicate_hooks/no_duplicate_hooks_test.go
package no_duplicate_hooks_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_duplicate_hooks"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoDuplicateHooksRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_duplicate_hooks.NoDuplicateHooksRule,
		[]rule_tester.ValidTestCase{
			{Code: `describe("foo", () => {
  beforeEach(() => {})
  test("bar", () => {
    someFn();
  })
})`},
			{Code: `beforeEach(() => {})
test("bar", () => {
  someFn();
})`},
			{Code: `describe("foo", () => {
  beforeAll(() => {}),
  beforeEach(() => {})
  afterEach(() => {})
  afterAll(() => {})

  test("bar", () => {
    someFn();
  })
})`},
			{Code: `describe.skip("foo", () => {
  beforeEach(() => {}),
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
})
describe("foo", () => {
  beforeEach(() => {}),
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
})`},
			{Code: `describe("foo", () => {
  beforeEach(() => {}),
  test("bar", () => {
    someFn();
  })
  describe("inner_foo", () => {
    beforeEach(() => {})
    test("inner bar", () => {
      someFn();
    })
  })
})`},
			{Code: `describe.each(['hello'])('%s', () => {
  beforeEach(() => {});

  it('is fine', () => {});
});`},
			{Code: `describe('something', () => {
  describe.each(['hello'])('%s', () => {
    beforeEach(() => {});

    it('is fine', () => {});
  });

  describe.each(['world'])('%s', () => {
    beforeEach(() => {});

    it('is fine', () => {});
  });
});`},
			{Code: "describe.each``('%s', () => {\n  beforeEach(() => {});\n\n  it('is fine', () => {});\n});"},
			{Code: "describe('something', () => {\n  describe.each``('%s', () => {\n    beforeEach(() => {});\n\n    it('is fine', () => {});\n  });\n\n  describe.each``('%s', () => {\n    beforeEach(() => {});\n\n    it('is fine', () => {});\n  });\n});"},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `describe("foo", () => {
  beforeEach(() => {}),
  beforeEach(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
				},
			},
			{
				Code: `describe.skip("foo", () => {
  beforeEach(() => {}),
  beforeAll(() => {}),
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 4, Column: 3},
				},
			},
			{
				Code: `describe.skip("foo", () => {
  afterEach(() => {}),
  afterEach(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
				},
			},
			{
				Code: `import { afterEach } from '@jest/globals';

describe.skip("foo", () => {
  afterEach(() => {}),
  afterEach(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 5, Column: 3},
				},
			},
			{
				Code: `import { afterEach, afterEach as somethingElse } from '@jest/globals';

describe.skip("foo", () => {
  afterEach(() => {}),
  somethingElse(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 5, Column: 3},
				},
			},
			{
				Code: `describe.skip("foo", () => {
  afterAll(() => {}),
  afterAll(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
				},
			},
			{
				Code: `afterAll(() => {}),
afterAll(() => {}),
test("bar", () => {
  someFn();
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 2, Column: 1},
				},
			},
			{
				Code: `describe("foo", () => {
  beforeEach(() => {}),
  beforeEach(() => {}),
  beforeEach(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
					{MessageId: "noDuplicateHook", Line: 4, Column: 3},
				},
			},
			{
				Code: `describe.skip("foo", () => {
  afterAll(() => {}),
  afterAll(() => {}),
  beforeAll(() => {}),
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
					{MessageId: "noDuplicateHook", Line: 5, Column: 3},
				},
			},
			{
				Code: `describe.skip("foo", () => {
  beforeEach(() => {}),
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
})
describe("foo", () => {
  beforeEach(() => {}),
  beforeEach(() => {}),
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 10, Column: 3},
				},
			},
			{
				Code: `describe("foo", () => {
  beforeAll(() => {}),
  test("bar", () => {
    someFn();
  })
  describe("inner_foo", () => {
    beforeEach(() => {})
    beforeEach(() => {})
    test("inner bar", () => {
      someFn();
    })
  })
})`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 8, Column: 5},
				},
			},
			{
				Code: `describe.each(['hello'])('%s', () => {
  beforeEach(() => {});
  beforeEach(() => {});

  it('is not fine', () => {});
});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
				},
			},
			{
				Code: `describe('something', () => {
  describe.each(['hello'])('%s', () => {
    beforeEach(() => {});

    it('is fine', () => {});
  });

  describe.each(['world'])('%s', () => {
    beforeEach(() => {});
    beforeEach(() => {});

    it('is not fine', () => {});
  });
});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 10, Column: 5},
				},
			},
			{
				Code: `describe('something', () => {
  describe.each(['hello'])('%s', () => {
    beforeEach(() => {});

    it('is fine', () => {});
  });

  describe.each(['world'])('%s', () => {
    describe('some more', () => {
      beforeEach(() => {});
      beforeEach(() => {});

      it('is not fine', () => {});
    });
  });
});`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 11, Column: 7},
				},
			},
			{
				Code: "describe.each``('%s', () => {\n  beforeEach(() => {});\n  beforeEach(() => {});\n\n  it('is fine', () => {});\n});",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 3, Column: 3},
				},
			},
			{
				Code: "describe('something', () => {\n  describe.each``('%s', () => {\n    beforeEach(() => {});\n\n    it('is fine', () => {});\n  });\n\n  describe.each``('%s', () => {\n    beforeEach(() => {});\n    beforeEach(() => {});\n\n    it('is not fine', () => {});\n  });\n});",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noDuplicateHook", Line: 10, Column: 5},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-duplicate-hooks.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-duplicate-hooks', {} as never, {
  valid: [
    {
      code: `
      describe("foo", () => {
        beforeEach(() => {})
        test("bar", () => {
          someFn();
        })
      })
    `,
    },
    {
      code: `
      beforeEach(() => {})
      test("bar", () => {
        someFn();
      })
    `,
    },
    {
      code: `
      describe("foo", () => {
        beforeAll(() => {}),
        beforeEach(() => {})
        afterEach(() => {})
        afterAll(() => {})

        test("bar", () => {
          someFn();
        })
      })
    `,
    },
    // multiple describe blocks
    {
      code: `
      describe.skip("foo", () => {
        beforeEach(() => {}),
        beforeAll(() => {}),
        test("bar", () => {
          someFn();
        })
      })
      describe("foo", () => {
        beforeEach(() => {}),
        beforeAll(() => {}),
        test("bar", () => {
          someFn();
        })
      })
    `,
    },
    // nested describe blocks
    {
      code: `
      describe("foo", () => {
        beforeEach(() => {}),
        test("bar", () => {
          someFn();
        })
        describe("inner_foo", () => {
          beforeEach(() => {})
          test("inner bar", () => {
            someFn();
          })
        })
      })
    `,
    },
    // describe.each blocks
    {
      code: `
      describe.each(['hello'])('%s', () => {
        beforeEach(() => {});

        it('is fine', () => {});
      });
    `,
    },
    {
      code: `
      describe('something', () => {
        describe.each(['hello'])('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });

        describe.each(['world'])('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });
      });
    `,
    },
    {
      code: `
      describe.each\`\`('%s', () => {
        beforeEach(() => {});

        it('is fine', () => {});
      });
    `,
    },
    {
      code: `
      describe('something', () => {
        describe.each\`\`('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });

        describe.each\`\`('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });
      });
    `,
    },
  ],
  invalid: [
    {
      code: `
        describe("foo", () => {
          beforeEach(() => {}),
          beforeEach(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 3,
          line: 3,
        },
      ],
    },
    {
      code: `
        describe.skip("foo", () => {
          beforeEach(() => {}),
          beforeAll(() => {}),
          beforeAll(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeAll' },
          column: 3,
          line: 4,
        },
      ],
    },
    {
      code: `
        describe.skip("foo", () => {
          afterEach(() => {}),
          afterEach(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'afterEach' },
          column: 3,
          line: 3,
        },
      ],
    },
    {
      code: `
        import { afterEach } from '@jest/globals';

        describe.skip("foo", () => {
          afterEach(() => {}),
          afterEach(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'afterEach' },
          column: 3,
          line: 5,
        },
      ],
    },
    {
      code: `
        import { afterEach, afterEach as somethingElse } from '@jest/globals';

        describe.skip("foo", () => {
          afterEach(() => {}),
          somethingElse(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'afterEach' },
          column: 3,
          line: 5,
        },
      ],
    },
    {
      code: `
        describe.skip("foo", () => {
          afterAll(() => {}),
          afterAll(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'afterAll' },
          column: 3,
          line: 3,
        },
      ],
    },
    {
      code: `
        afterAll(() => {}),
        afterAll(() => {}),
        test("bar", () => {
          someFn();
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'afterAll' },
          column: 1,
          line: 2,
        },
      ],
    },
    {
      code: `
        describe("foo", () => {
          beforeEach(() => {}),
          beforeEach(() => {}),
          beforeEach(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 3,
          line: 3,
        },
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 3,
          line: 4,
        },
      ],
    },
    {
      code: `
        describe.skip("foo", () => {
          afterAll(() => {}),
          afterAll(() => {}),
          beforeAll(() => {}),
          beforeAll(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'afterAll' },
          column: 3,
          line: 3,
        },
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeAll' },
          column: 3,
          line: 5,
        },
      ],
    },
    // multiple describe blocks
    {
      code: `
        describe.skip("foo", () => {
          beforeEach(() => {}),
          beforeAll(() => {}),
          test("bar", () => {
            someFn();
          })
        })
        describe("foo", () => {
          beforeEach(() => {}),
          beforeEach(() => {}),
          beforeAll(() => {}),
          test("bar", () => {
            someFn();
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 3,
          line: 10,
        },
      ],
    },
    // nested describe blocks
    {
      code: `
        describe("foo", () => {
          beforeAll(() => {}),
          test("bar", () => {
            someFn();
          })
          describe("inner_foo", () => {
            beforeEach(() => {})
            beforeEach(() => {})
            test("inner bar", () => {
              someFn();
            })
          })
        })
      `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 5,
          line: 8,
        },
      ],
    },
    // describe.each blocks
    {
      code: `
      describe.each(['hello'])('%s', () => {
        beforeEach(() => {});
        beforeEach(() => {});

        it('is not fine', () => {});
      });
    `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 3,
          line: 3,
        },
      ],
    },
    {
      code: `
      describe('something', () => {
        describe.each(['hello'])('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });

        describe.each(['world'])('%s', () => {
          beforeEach(() => {});
          beforeEach(() => {});

          it('is not fine', () => {});
        });
      });
    `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 5,
          line: 10,
        },
      ],
    },
    {
      code: `
      describe('something', () => {
        describe.each(['hello'])('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });

        describe.each(['world'])('%s', () => {
          describe('some more', () => {
            beforeEach(() => {});
            beforeEach(() => {});

            it('is not fine', () => {});
          });
        });
      });
    `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 7,
          line: 11,
        },
      ],
    },
    {
      code: `
      describe.each\`\`('%s', () => {
        beforeEach(() => {});
        beforeEach(() => {});

        it('is fine', () => {});
      });
    `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 3,
          line: 3,
        },
      ],
    },
    {
      code: `
      describe('something', () => {
        describe.each\`\`('%s', () => {
          beforeEach(() => {});

          it('is fine', () => {});
        });

        describe.each\`\`('%s', () => {
          beforeEach(() => {});
          beforeEach(() => {});

          it('is not fine', () => {});
        });
      });
    `,
      errors: [
        {
          messageId: 'noDuplicateHook',
          data: { hook: 'beforeEach' },
          column: 5,
          line: 10,
        },
      ],
    },
  ],
});
